src/mlpack/core/metrics/bleu_impl.hpp

```diff
@@ -0,0 +1,200 @@
+ElemType BLEU<ElemType, PrecisionType>::Evaluate(
+    const ReferenceCorpusType& referenceCorpus,
+    const TranslationCorpusType& translationCorpus,
+    const bool smooth)
+{
+  // WordVector is a string container type.
+  // Also, TranslationCorpusType is an array of such containers.
+  typedef typename TranslationCorpusType::value_type WordVector;
+
+  // matchesByOrder: It catches how many times sequence of a particular order
+  // is encountered in both reference corpus and translation corpus.
+  std::vector<size_t> matchesByOrder(maxOrder, 0);
+
+  // possibleMatchesByOrder: It tracks how many possible matches can be in the
+  // translation corpus.
+  std::vector<size_t> possibleMatchesByOrder(maxOrder, 0);
+
+  // referenceLength: It is the sum of minimum length of the paragraph from
+  // various documents.
+  // translationLength: It is the sum of length of each paragraphs.
+  referenceLength = 0, translationLength = 0;
+
+  auto refIt = referenceCorpus.cbegin();
+  auto trIt = translationCorpus.cbegin();
+  for (; refIt != referenceCorpus.cend(), trIt != translationCorpus.cend();
+      ++refIt, ++trIt)
+  {
+    size_t min = std::numeric_limits<size_t>::max();
+    for (const auto& t : *refIt)
+    {
+      if (min > t.size())
+      {
+        min = t.size();
+      }
+    }
+
+    if (min == std::numeric_limits<size_t>::max())
+      min = 0;
+
+    referenceLength += min;
+    translationLength += trIt->size();
+
+    // mergedRefNGramCounts: It accumulates all the similar n-grams from
+    // various references or documents, so that there is no repetition of
+    // any key (sequence of order n).
+    std::map<WordVector, size_t> mergedRefNGramCounts;
+    for (const auto& t : *refIt)
+    {
+      // ngram: It holds the n-grams of each document/reference.
+      const std::map<WordVector, size_t> ngrams = GetNGrams(t);
+      for (auto it = ngrams.cbegin(); it != ngrams.cend(); ++it)
+      {
+        mergedRefNGramCounts[it->first] = std::max(it->second,
+            mergedRefNGramCounts[it->first]);
+      }
+    }
+    // translationNGramCounts: It extracts the n-grams of the generated text
+    // sequence.
+    const std::map<WordVector, size_t> translationNGramCounts
+        = GetNGrams(*trIt);
+
+    // overlap: It holds those key (sequence of order n) which are common to
+    // reference corpus and translation corpus.
+    std::map<WordVector, size_t> overlap;
+    for (auto it = translationNGramCounts.cbegin();
+         it != translationNGramCounts.cend();
+         ++it)
+    {
+      if (mergedRefNGramCounts.find(it->first) != mergedRefNGramCounts.end())
+      {
+        // If the key (sequence of order n) is present in both translation
+        // corpus as well as reference corpus, then the minimum number of
+        // counts it has occurred in any is considered.
+        overlap[it->first] = std::min(mergedRefNGramCounts.count(it->first),
+                                      it->second);
+      }
```

Comment:
I think that's incorrect since `mergedRefNGramCounts.count(it->first)` always equals to `1`. If I remember right I proposed the following implementation
```suggestion
      auto mergedIt = mergedRefNGramCounts.find(it->first);
      if (mergedIt != mergedRefNGramCounts.end())
      {
        // If the key (sequence of order n) is present in both translation
        // corpus as well as reference corpus, then the minimum number of
        // counts it has occurred in any is considered.
        overlap[it->first] = std::min(mergedIt->second, it->second);
      }
```
Why didn't you apply it? I think it's correct.

Reply:
Ohh...there was no reason, I was just experimenting them 😅 and got busy in fixing errors in other PRs. My bad.